src/player.cpp

```diff
@@ -5467,169 +5501,243 @@ void player::update_needs()
             add_effect("lack_sleep", 50);
         }
     }
+}
 
-    if( calendar::once_every(MINUTES(5)) ) { // Hunger, thirst, & fatigue up every 5 minutes
-        if( (!has_trait("LIGHTEATER") || !one_in(3)) &&
-            (!has_bionic("bio_recycler") || calendar::once_every(MINUTES(30)) ) &&
-            !(has_trait("DEBUG_LS") )) {
-            mod_hunger(1);
-            if (has_trait("HUNGER")) {
-                if (one_in(2)) {
-                    mod_hunger(1);
-                }
-            }
-            if (has_trait("MET_RAT")) {
-                if (!one_in(3)) {
-                    mod_hunger(1);
-                }
-            }
-            if (has_trait("HUNGER2")) {
-                mod_hunger(1);
-            }
-            if (has_trait("HUNGER3")) {
-                mod_hunger(2);
-            }
-        }
-        if( (!has_bionic("bio_recycler") || calendar::once_every(MINUTES(10)) ) &&
-            (!has_trait("PLANTSKIN") || !one_in(5)) &&
-            (!has_trait("DEBUG_LS")) ) {
-            thirst++;
-            if (has_trait("THIRST")) {
-                if (one_in(2)) {
-                    thirst++;
-                }
-            }
-            if (has_trait("THIRST2")) {
-                thirst++;
-            }
-            if (has_trait("THIRST3")) {
-                thirst += 2;
-            }
-        }
-        // Sanity check for negative fatigue value.
-        if (fatigue < -1000) {
-            fatigue = -1000;
-        }
+void player::update_needs( int rate_multiplier )
+{
+    // Hunger, thirst, & fatigue up every 5 minutes
+    const bool debug_ls = has_trait("DEBUG_LS");
+    const bool has_recycler = has_bionic("bio_recycler");
+    const bool asleep = in_sleep_state();
+    const bool hibernating = asleep && is_hibernating();
+    float hunger_rate = 1.0f;
```

Comment:
Not a must-have, particularly for this PR, but it'd be nice to extract these guys to player::hunger_rate(), plater::fatigue_rate(), etc.